numpy-stubs/__init__.pyi

```diff
@@ -792,3 +797,124 @@ class AxisError(ValueError, IndexError):
+_ScalarGeneric = TypeVar("_ScalarGeneric", bound=generic)
+_ScalarBuiltin = Union[str, bytes, dt.date, dt.timedelta, bool, int, float, complex]
+_Scalar = Union[_ScalarBuiltin, generic]
+
+# An array-like object consisting of integers
+_Int = Union[int, integer]
+_ArrayLikeIntNested = Any  # TODO: wait for support for recursive types
+_ArrayLikeInt = Union[
+    _Int, ndarray, Sequence[_Int], Sequence[_ArrayLikeIntNested]  # TODO: ndarray[int]
+]
+
+# An array-like object consisting of strings
+_ArrayLikeStr = Union[
+    str, str_, Sequence[Union[str, str_]], ndarray
+]  # TODO: ndarray[str]
+
+# The signature of take() follows a common theme with its overloads:
+# 1. A generic comes in; the same generic comes out
+# 2. A scalar comes in; a generic comes out
+# 3. An array-like object comes in; some keyword ensures that a generic comes out
+# 4. An array-like object comes in; an ndarray or generic comes out
+@overload
+def take(
+    a: _ScalarGeneric,
+    indices: int,
+    axis: Optional[int] = ...,
+    out: Optional[ndarray] = ...,
+    mode: _Mode = ...,
+) -> _ScalarGeneric: ...
+@overload
+def take(
+    a: _Scalar,
+    indices: int,
+    axis: Optional[int] = ...,
+    out: Optional[ndarray] = ...,
+    mode: _Mode = ...,
+) -> generic: ...
+@overload
+def take(
+    a: _ArrayLike,
+    indices: int,
+    axis: Optional[int] = ...,
+    out: Optional[ndarray] = ...,
+    mode: _Mode = ...,
+) -> generic: ...
+@overload
+def take(
+    a: _ArrayLike,
+    indices: _ArrayLikeInt,
+    axis: Optional[int] = ...,
+    out: Optional[ndarray] = ...,
+    mode: _Mode = ...,
+) -> Union[generic, ndarray]: ...
+def reshape(a: _ArrayLike, newshape: _ShapeLike, order: _Order = ...) -> ndarray: ...
+@overload
+def choose(
+    a: _ScalarGeneric,
+    choices: Union[Sequence[_ArrayLike], ndarray],  # TODO: ndarray[_ArrayLike]
+    out: Optional[ndarray] = ...,
+    mode: _Mode = ...,
+) -> _ScalarGeneric: ...
+@overload
+def choose(
+    a: _Scalar,
+    choices: Union[Sequence[_ArrayLike], ndarray],  # TODO: ndarray[_ArrayLike]
+    out: Optional[ndarray] = ...,
+    mode: _Mode = ...,
+) -> generic: ...
+@overload
+def choose(
+    a: _ArrayLike,
+    choices: Union[Sequence[_ArrayLike], ndarray],  # TODO: ndarray[_ArrayLike]
+    out: Optional[ndarray] = ...,
+    mode: _Mode = ...,
+) -> ndarray: ...
+def repeat(
+    a: _ArrayLike, repeats: _ArrayLikeInt, axis: Optional[int] = ...
+) -> ndarray: ...
+def put(a: ndarray, ind: _ArrayLikeInt, v: _ArrayLike, mode: _Mode = ...) -> None: ...
+def swapaxes(
+    a: Union[Sequence[_ArrayLike], ndarray],  # TODO: ndarray[_ArrayLike]
+    axis1: int,
+    axis2: int,
+) -> ndarray: ...
+def transpose(
+    a: _ArrayLike, axes: Union[None, Sequence[int], ndarray] = ...  # TODO: ndarray[int]
+) -> ndarray: ...
+def partition(
+    a: _ArrayLike,
+    kth: _ArrayLikeInt,
+    axis: Optional[int] = ...,
+    kind: _PartitionKind = ...,
+    order: Optional[_ArrayLikeStr] = ...,
```

Comment:
This appears to be an instance where we don't want `ndarray`:

```python
>>> x = np.zeros((10,), dtype=[('foo', np.float64), ('bar', np.float64)])
>>> np.partition(x, kth=0, order=['foo', 'bar'])
array([(0., 0.), (0., 0.), (0., 0.), (0., 0.), (0., 0.), (0., 0.),
       (0., 0.), (0., 0.), (0., 0.), (0., 0.)],
      dtype=[('foo', '<f8'), ('bar', '<f8')])
>>> np.partition(x, kth=0, order=('foo', 'bar'))
array([(0., 0.), (0., 0.), (0., 0.), (0., 0.), (0., 0.), (0., 0.),
       (0., 0.), (0., 0.), (0., 0.), (0., 0.)],
      dtype=[('foo', '<f8'), ('bar', '<f8')])
>>> np.partition(x, kth=0, order=np.array(['foo', 'bar']))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 6, in partition
  File "/Users/Josh/Projects/numpy-stubs-dev/venv/lib/python3.7/site-packages/numpy/core/fromnumeric.py", line 746, in partition
    a.partition(kth, axis=axis, kind=kind, order=order)
  File "/Users/Josh/Projects/numpy-stubs-dev/venv/lib/python3.7/site-packages/numpy/core/_internal.py", line 383, in _newnames
    raise ValueError("unsupported order value: %s" % (order,))
ValueError: unsupported order value: ['foo' 'bar']
```

So I think here and below should be `Optional[Union[str, Sequence[str]]`.

Reply:
Good catch; an easy fix to implement.